internal/component/database_observability/mysql/collector/schema_table_test.go

```
@@ -121,6 +121,233 @@ func TestSchemaTable(t *testing.T) {
+			WillReturnRows(
+				sqlmock.NewRows([]string{
+					"table_name",
+					"create_statement",
+				}).AddRow(
+					"some_schema.some_table",
+					"CREATE TABLE some_table (id INT)",
+				),
+			)
+
+		mock.ExpectQuery(selectColumnNames).WithArgs("some_schema", "some_table").RowsWillBeClosed().
+			WillReturnRows(
+				sqlmock.NewRows([]string{
+					"COLUMN_NAME",
+					"COLUMN_DEFAULT",
+					"IS_NULLABLE",
+					"COLUMN_TYPE",
+					"COLUMN_KEY",
+					"EXTRA",
+				}).AddRow(
+					"id",
+					"null",
+					"NO",
+					"int",
+					"PRI",
+					"auto_increment",
+				),
+			)
+
+		err = collector.Start(context.Background())
+		require.NoError(t, err)
+
+		require.Eventually(t, func() bool {
+			return len(lokiClient.Received()) == 3
+		}, 5*time.Second, 100*time.Millisecond)
+
+		collector.Stop()
+		lokiClient.Stop()
+
+		require.Eventually(t, func() bool {
+			return collector.Stopped()
+		}, 5*time.Second, 100*time.Millisecond)
+
+		err = mock.ExpectationsWereMet()
+		require.NoError(t, err)
+
+		lokiEntries := lokiClient.Received()
+		for _, entry := range lokiEntries {
+			require.Equal(t, model.LabelSet{"job": database_observability.JobName}, entry.Labels)
+		}
+		require.Equal(t, `level=info msg="schema detected" op="schema_detection" instance="mysql-db" schema="some_schema"`, lokiEntries[0].Line)
+		require.Equal(t, `level=info msg="table detected" op="table_detection" instance="mysql-db" schema="some_schema" table="some_table"`, lokiEntries[1].Line)
+		require.Equal(t, fmt.Sprintf(`level=info msg="create table" op="create_statement" instance="mysql-db" schema="some_schema" table="some_table" create_statement="%s" table_spec="%s"`, base64.StdEncoding.EncodeToString([]byte("CREATE TABLE some_table (id INT)")), base64.StdEncoding.EncodeToString([]byte(`{"columns":[{"name":"id","type":"int","not_null":true,"auto_increment":true,"primary_key":true,"default_value":"null"}]}`))), lokiEntries[2].Line)
```

I didn't understand how this test doesn't read from cache like the other test does. How are the two tests different? Is it just that the other test waits until the number of lokiEntries double to 6?

---

This test only writes to cache, because it loops only once. The next test writes in the first loop and then reads in the second loop. I could use a suggestion for better naming.

---

Does it do a second in the next test only because of the following or did I miss anything else?
```
		require.Eventually(t, func() bool {
			return len(lokiClient.Received()) == 6
		}, 50000*time.Second, 100*time.Millisecond)
```

---

Thanks to your comment I noticed I had left `50000` as I was debugging the test.

> Does it do a second loop in the next test only because of the following

It loops twice because the mock database is instructed to expect/return values for this sequence of queries:

- selectSchemaName (first loop)
- selectTableName (first loop)
- SHOW CREATE TABLE (first loop)
- selectColumnNames (first loop)
- selectSchemaName (second loop)
- selectTableName (second loop)

Note that in the second loop we expect fewer queries because that content must be cached.

Other tests stop after the first loop because the mock database return an error after all expectations are met.

---

Ah ok thanks for the explanation. 

Hmm I don't have a great idea for improving the names. Maybe some long explicit test names if you want. "detects table schema, writes to cache, then retrieves from cache"

In this case where it's a little complicated, I would maybe just include what you explained here as comments.
Maybe group the (first loop) mocks together and above a comment like `// first loop, cache is written` then group the (second loop) mocks together and above a comment like `// second loop, note that show create table and select column names are not called: they were retrieved from the cache`. It might not hurt to have these comments if we consider these tests to be a bit complicated. (I am usually more in favor of self-explanatory code, but in this case it might make figuring them out a lot easier in the future for a newcomer or while debugging a problem.)

By the way, I don't know if this would add value or make anything clearer, but are you able make assertions on the cache directly itself? (collector.cache) 

Anyway just my 2c, I defer to your best judgement.

---

> By the way, I don't know if this would add value or make anything clearer, but are you able make assertions on the cache directly itself? (collector.cache)

One thing we should add later on is debug metrics (those exposed by Alloy itself), which we could use for tests assertions. Right now it'd be a bit tedious to assert it in tests though.

---

Sorry I might've meant a different thing.
I meant that in the test, I saw I can access the methods of `collector.cache`. And you're putting stuff into the cache, right? For example, would it be possible to assert that collector.cache has some key or some values that exist that you expect to be there? (I think it is less end-to-end than the lokiEntries, so disregard this if it doesn't add value. I don't mean to beat a dead horse or anything.)

---

I see, pushed one more commit with a new assertion. Not a big fan of testing it via accessing a private field, but don't want to expose the cache with public methods either.